### PR TITLE
feat(discord): add change online status on stream live and event listener

### PR DESCRIPTION
**Pull request by liadala**
added discord status change depends on stream status
added toogle for this feature
added dropdown for default bot status
###### CHECKLIST

- [ ] tests are changed or added
- [ ] documentation is changed or added
- [ ] locales are changed or added
- [ ] db relationship (tools/database) are changed or added
- [ ] commit message follows [commit guidelines](https://github.com/sogehige/sogeBot/blob/master/CONTRIBUTING.md#commit-guidelines)

### CODE REVIEW

File: src/bot/integrations/discord.ts [hunk 1]
```diff
@@ -98,6 +98,20 @@ class Discord extends Integration {
   })
   sendGeneralAnnounceToChannel = '';
 
+  @settings('bot')
+  @ui({
+    type: 'selector', values: ['online', 'invisible', 'dnd'],
+    if: () => self.guild.length > 0,
+  })
+  onlinePresence: 'online' | 'idle' | 'invisible' | 'dnd' = 'online';
```
Comment (sogehige) [inline src/bot/integrations/discord.ts:106]: https://github.com/sogehige/sogeBot/pull/4030#discussion_r470027846
https://github.com/sogehige/sogeBot/pull/4030#discussion_r470028519

Comment (liadala) [reply src/bot/integrations/discord.ts:106]: > also this seem to be confusing to me as well...we should have configurable online presense for offline/online stream states. Also we should have configurable activity as well (maybe just toggle to set streaming activity during online stream) -> thus keeping old behavior

if toggle is enabled (default) bot change its status to streaming if stream goes live
if togge is disabled bot stay in selected status like bot actualy do

> so we should have
this.onlinePresence -> presence when stream is online
this.offlinePresence -> presence when stream is offline

i will add this if i can

and i request onHosting() -> set bot to watching status
as a third option

**Comment by liadala** [comment]
changed to have 2 selectors for default satus and for onStream status

File: src/bot/integrations/discord.ts [hunk 2]
```diff
@@ -98,6 +98,28 @@ class Discord extends Integration {
...
+    if: () => self.guild.length > 0,
+  })
+  togglePresenceOnStream = true;
+ */
+  @settings('bot')
+  @ui({
+    type: 'selector', values: ['streaming', 'online', 'idle', 'invisible', 'dnd'],
+    if: () => self.guild.length > 0,
+  })
+  togglePresenceOnStream: 'streaming' | 'online' | 'idle' | 'invisible' | 'dnd' = 'online';
```
Comment (sogehige) [inline src/bot/integrations/discord.ts:121]: are you sure this is doing exactly what you want? I mean `toggle` should be true/false. Maybe change naming convention.

File: src/bot/integrations/discord.ts [hunk 3]
```diff
@@ -371,6 +394,28 @@ class Discord extends Integration {
...
+  @onChange('onlinePresence')
+  async changeClientOnlinePresence() {
+    try {
+      if (api.isStreamOnline) {
+        if (this.togglePresenceOnStream === 'streaming') {
+          this.client?.user?.setStatus('online');
+          this.client?.user?.setPresence({ status: 'online', activity: { name: `${api.stats.currentTitle}`, type: 'STREAMING', url: `https://twitch.tv/${oauth.generalChannel}`} });
+        } else {
+          this.client?.user?.setStatus(this.togglePresenceOnStream);
+          this.client?.user?.setPresence({ status: this.togglePresenceOnStream, activity: { name: '', type: undefined, url: '' } });
```
Comment (sogehige) [inline src/bot/integrations/discord.ts:410]: Also I am not super happy with empty activity, can it be customized by settings? what can be in name? type? url? We can probably add more variables to this, couldn't we?

Comment (liadala) [reply src/bot/integrations/discord.ts:410]: name can be any string
type [ActivityType](https://discord.js.org/#/docs/main/stable/typedef/ActivityType)
url is twitch link ONLY

Comment (sogehige) [reply src/bot/integrations/discord.ts:410]: if activity can be any string, we can have it configurable same as status. if you use 

```
const activityString = await new Message(this.offlineActivityName).parse({});
```

than we can use all filters

Comment (liadala) [reply src/bot/integrations/discord.ts:410]: `the setPresence( activity{name: activityString} )`
can only exist if type is set to streaming/watching/listening
a discord bot cant use custom_status this is limited to normal user only
so its limited in usecase

but having online activity name config is a idea
and a toggle for onHost --> `setPresence( activity{name: $host type: 'WATCHING'} )`

File: src/bot/integrations/discord.ts [hunk 4]
```diff
@@ -98,6 +98,28 @@ class Discord extends Integration {
...
   sendGeneralAnnounceToChannel = '';
 
+  @settings('bot')
+  @ui({
+    type: 'selector', values: ['online', 'idle', 'invisible', 'dnd'],
+    if: () => self.guild.length > 0,
+  })
+  onlinePresence: 'online' | 'idle' | 'invisible' | 'dnd' = 'online';
+
+  /*
```
Comment (sogehige) [inline src/bot/integrations/discord.ts:108]: this is commented

Comment (liadala) [reply src/bot/integrations/discord.ts:108]: i know i remove it in next commit also have regrouped the status to its own label

**Comment by liadala** [comment]
moved status settings into its own label
and added text fields for custom text

File: locales/en/ui/integrations/discord.json [hunk 5]
```diff
@@ -13,6 +13,11 @@
     "noChannelSelected": "no channel selected",
     "noRoleSelected": "no role selected",
     "noGuildSelected": "no guild selected",
-    "noGuildSelectedBox": "Select guild where bot should work and you'll see more settings"
+    "noGuildSelectedBox": "Select guild where bot should work and you'll see more settings",
+    "onlinePresence": "Bot Default Status",
```
Comment (sogehige) [inline locales/en/ui/integrations/discord.json:17]: unused locale

File: src/bot/integrations/discord.ts [hunk 6]
```diff
@@ -371,6 +402,42 @@ class Discord extends Integration {
     } catch (e) {
       warning(e.stack);
     }
+
+    this.changeClientOnlinePresence();
```
Comment (sogehige) [inline src/bot/integrations/discord.ts:406]: can you move this to first line? (to be consistent with updateStreamStartAnnounce())

File: locales/en/ui/integrations/discord.json (hunk 5 above)
Comment (liadala) [reply locales/en/ui/integrations/discord.json]: removed

File: src/bot/integrations/discord.ts (hunk 6 above)
Comment (liadala) [reply src/bot/integrations/discord.ts]: Changed with [729bcf5](https://github.com/liadala/sogeBot/commit/729bcf5447f1bc815ddd566aeeaa192117855f40)

File: src/bot/integrations/discord.ts [hunk 7]
```diff
@@ -373,6 +410,67 @@ class Discord extends Integration {
     }
   }
 
+  @onChange('onlinePresenceStatusOnStreamName')
+  @onChange('onlinePresenceStatusDefaultName')
+  @onChange('onlinePresenceStatusOnStream')
+  @onChange('onlinePresenceStatusDefault')
+  async changeClientOnlinePresence() {
```
Comment (sogehige) [inline src/bot/integrations/discord.ts:417]: you can add @onStreamStart() and @onStreamEnd() here as well next time, but that's matter of preference

File: src/bot/integrations/discord.ts [hunk 8]
```diff
@@ -373,6 +410,67 @@ class Discord extends Integration {
...
+          this.client?.user?.setActivity('');
+          this.client?.user?.setStatus(this.onlinePresenceStatusDefault);
+        }
+      }
+    } catch (e) {
+      warning(e.stack);
+    }
+  }
+
+  public async sendDM(text: string, dMchannel: string): Promise<any> {
```
Comment (sogehige) [inline src/bot/integrations/discord.ts:447]:
```suggestion
  public async sendDM(text: string, dMchannel: string): Promise<void> {
```

**Pull request merged**